asf_heat_pump_suitability/pipeline/prepare_features/grid_capacity.py

```diff
@@ -0,0 +1,316 @@
+
+    if isinstance(value, (int, float)):
+        return float(value)
+
+    value = str(value).strip().lower()
+
+    # Pattern to match "X x Y" format (e.g., "2 x 12.5")
+    pattern = r"(\d+(?:\.\d+)?)\s*x\s*(\d+(?:\.\d+)?)"
+    match = re.match(pattern, value)
+
+    if match:
+        count = float(match.group(1))
+        individual_capacity = float(match.group(2))
+        return count * individual_capacity
+
+    try:
+        return float(value)
+    except ValueError:
+        logging.warning(f"Could not parse capacity value: {value}")
+        return np.nan
+
+
+def generate_substations_gdf() -> gpd.GeoDataFrame:
+    """
+    Generate a combined GeoDataFrame containing all primary substations and their service areas
+    across Distribution Network Operators (DNOs).
+
+    Combines data from:
+        - Electricity North West (ENW)
+        - Northern Powergrid (NPg)
+        - Scottish Power Energy Networks (SPEN)
+        - Scottish and Southern Electricity Networks (SSEN)
+        - UK Power Networks (UKPN)
+        - Western Power Distribution (WPD)
+
+    Returns:
+        gpd.GeoDataFrame: DataFrame with EPSG:4326 (WGS84) projection containing:
+            - id: Unique substation identifier (format varies by operator)
+            - firm_capacity_mva: Maximum power the substation can safely deliver (MVA)
+            - peak_demand_mva: Maximum observed power demand at the substation (MVA)
+            - geo_shape: Polygon geometry representing the substation's service area
+            - operator: Distribution network operator code
+                       (one of: "ENW", "NPg", "SPEN", "SSEN", "UKPN", "WPD")
+    """
+    return pd.concat(
+        [
+            generate_enw_gdf(),
+            generate_npg_gdf(),
+            generate_spen_gdf(),
+            generate_ssen_gdf(),
+            generate_ukpn_gdf(),
+            generate_wpd_gdf(),
+        ]
+    )
+
+
+def distribute_substation_headroom(
+    substations_gdf: gpd.GeoDataFrame,
+    lsoa_gdf: gpd.GeoDataFrame,
+    households_df: pd.DataFrame,
+) -> pd.DataFrame:
+    """
+    Distribute substation headroom to LSOAs, weighted by household count.
+
+    Args:
+        substations_gdf: GeoDataFrame of substations with headroom data.
+        lsoa_gdf: GeoDataFrame of LSOA boundaries.
+        households_df: DataFrame with household counts per LSOA.
+
+    Returns:
+        DataFrame with distributed headroom per LSOA.
+    """
+    # Merge LSOA boundaries with household counts
+    lsoa_with_households = lsoa_gdf.merge(households_df, on="LSOA21CD", how="left")
+
+    # Perform spatial join between LSOAs and substations
+    joined = gpd.sjoin(
+        lsoa_with_households, substations_gdf, how="inner", predicate="intersects"
+    ).rename(columns={"index_right": "substation_id"})
+
+    # Calculate total households served by each substation
+    substation_total_households = (
+        joined.groupby("substation_id")["household_count"].sum().reset_index()
```

Comment:
Should this use the existing `id` column from the `substations_gdf` instead of the index? I guess it should be the same and maybe it's better (if there are null ids). Just wanted to check though.